internal/stores/hostdb.go

```diff
@@ -151,23 +252,72 @@ func (db *SQLStore) Hosts(notSince time.Time, max int) ([]hostdb.Host, error) {
 	return hosts, err
 }
 
+func hostByPubKey(tx *gorm.DB, hostKey consensus.PublicKey) (dbHost, error) {
+	var h dbHost
+	err := tx.Where("public_key", gobEncode(hostKey)).
+		Take(&h).Error
+	return h, err
+}
+
 // RecordInteraction records an interaction with a host. If the host is not in
 // the store, a new entry is created for it.
-func (db *SQLStore) RecordInteraction(hostKey consensus.PublicKey, hi hostdb.Interaction) error {
+func (db *SQLStore) RecordHostInteractions(hostKey consensus.PublicKey, interactions []hostdb.Interaction) error {
+	dbInteractions := make([]dbInteraction, len(interactions))
+	var successful, failed float64
+
+	var scanQuery string
+	var scanArgs []interface{}
+	for i, interaction := range interactions {
+		dbInteractions[i] = dbInteraction{
+			Result:    interaction.Result,
+			Success:   interaction.Success,
+			Timestamp: interaction.Timestamp.UTC(),
+			Type:      interaction.Type,
+		}
+		if interaction.Success {
+			successful++
+		} else {
+			failed++
+		}
+		// Handle scan edge case.
+		if interaction.Type == hostdb.InteractionTypeScan {
+			timeToAddTo := "uptime"
+			if !interaction.Success {
+				timeToAddTo = "downtime"
+			}
+			var sr hostdb.ScanResult
+			if interaction.Success {
+				if err := json.Unmarshal(interaction.Result, &sr); err != nil {
+					return err
+				}
+			}
+			scanQuery += fmt.Sprintf(`UPDATE hosts
+			              SET total_scans = total_scans + 1,
+				      previous_scan_success = last_scan_success,
+				      last_scan_success = ?,
+				      %s = CASE WHEN last_scan == 0 THEN last_scan ELSE ? - last_scan END,
```

Comment:
if `last_scan` is uninitiated, will this cause a huge uptime after the first successful scan?